lib/ddtrace/span.rb

```diff
@@ -160,6 +165,36 @@ def set_error(e)
       set_tag(Ext::Errors::STACK, e.backtrace) unless e.backtrace.empty?
     end
 
+    # Mark the span started at the current time.
+    def start(start_time = nil)
```

Comment:
The API change here is:

`Span#start()` which will optionally take a unix timestamp as the start time.

If a start time is provided, then we do not use a monotonic clock for tracking the duration.

`Span#finish()` which we already had and still optionally takes a unix timestamp as the end time.

If a finish time is provided, then we do not use the monotonic duration we may have started for the duration.

When trying to determine the duration of the span, if either start or finish time were provided, we just subtract the unix timestamps, otherwise we use the monotonic duration?

Reply:
That's correct. 

`Span.start` sets both `@start_time` (wall clock time) and `@duration_start` (monotonic time)
`Span.finish` sets both `@end_time` (wall clock time) and `@duration_end` (monotonic time)

Both methods can toggle a flag `@wall_clock_duration` which determines whether to use wall clock time or monotonic time. This flag is set based on whether a start or end time is provided to either method.

In the event `@wall_clock_duration` is set to true at any point when starting or finishing the span, these methods also defensively update `@duration_start` and `@duration_end` to represent wall clock time.

There is an additional change to the `Span` API, which is the addition of a `Span.duration` method, which returns either wall time or monotonic time depending on the `@wall_clock_duration` flag.

The `span.duration` is work we'd been doing already when serialising the span, it's just now formalised in a method on the Span class so that we don't have to copy that logic everywhere we want to serialize the Span.

@marcotc what do you think about these changes? This removes the need for a separate `Duration` class as well as the use of `Forwardable` which i know we had some memory perf concerns about.

Reply:
I keep thinking about how this logic can be simplified, it _feels_ just a hair more complicated than it needs to be, although the behavior is correct.


```ruby
def start(start_time=nil)
    return if started?
    @start_time = start_time || Time.now.utc
    @duration_start = start_time.nil? duration_marker : nil
end

def finish(finish_time=nil)
    return if finished?

    now = Time.now.utc
    start(now) unless started?
    @end_time = finish_time || now
    @duration_end = finish_time.nil? ? duration_marker : nil
end

def duration
    # We are missing either the start or end monotonic clock, compute duration from unix timestamps
    if @duration_start.nil? || @duration_end.nil?
        (@end_time - @start_time).to_f rescue 0.0
    else
        (@duration_end - @duration_start).to_f rescue 0.0
    end
end
```


If we can use `@duration_start.nil?` or `@duration_end.nil?` as the `@wall_clock_duration` then we can probably simplify this a bit.

And if that is the case, using that to determine if duration uses `@duration_{start,end}` or `@{start,end}_time`, then we shouldn't need to worry cases when `@duration_{start,end}` could be a unix timestamp or a monotonic, it means it'll always be monotonic or `nil`.


What do you think? (I am not positive this actually works btw, just been trying to think how to simplify and best representation I could get)

Reply:
I've already cleaned this up a bit, you might be reviewing stale code

Reply:
oh, hm, yea i think those changes are reasonable and remove the need for a flag, let me see if it works

Reply:
No, think my suggestion still stands.

The piece that bothers me a bit is the fact that `@duration_{start,end}` could be either `nil`, a unix timestamp, or a monotonic clock.

This means it'll likely be more prone to coding errors, if somehow `@duration_start` is monotonic and `@duration_end` is unix timestamp, then duration will be way off.

Reply:
Just to be clear, `@duration_start` and `@duration_end` are not publicly accessible so, the coding errors would be our own internal errors. Which is certainly possible, so i agree it's good to simplify.